myst_nbs/case_studies/regression_discontinuity.myst.md

```diff
@@ -0,0 +1,219 @@
+    ax.axvline(x=threshold, color="k", ls="--", lw=3, label="treatment threshold")
+    ax.set(xlabel=r"observed $x$ (pre-test)", ylabel=r"observed $y$ (post-test)")
+    plt.legend()
+    return ax
+
+
+plot_data(df);
+```
+
++++ {"tags": []}
+
+## Sharp regression discontinuity model
+
+We can define our Bayesian regression discontinuity model as:
+
+$$
+\begin{aligned}
+\Delta & \sim \text{Cauchy}(0, 1) \\
+\sigma & \sim \text{HalfNormal}(0, 1) \\
+\mu & = x_i + \underbrace{\Delta \cdot treated_i}_{\text{treatment effect}} \\
+y_i & \sim \text{Normal}(\mu, \sigma)
+\end{aligned}
+$$
+
+where:
+- $\Delta$ is the size of the discontinuity, 
+- $\sigma$ is the standard deviation of change in the pre- to post-test scores,
+- $x_i$ and $y_i$ are observed pre- and post-test measures for unit $i$, and 
+- $treated_i$ is an observed indicator variable (0 for control group, 1 for test group).
+
+Notes:
+- If the pre-test ($x$) and post-test ($y$) measures where not the same, then we would want to add additional slope and intercept parameters to this model. But because we assume the same measure is being taken pre- and post-test, then we do not need this.
+- We assume we have accurately observed whether a unit has been treated or not. That is, this model assumes a sharp discontinuity with no uncertainty.
+
+```{code-cell} ipython3
+with pm.Model() as model:
+    x = pm.MutableData("x", df.x, dims="obs_id")
+    treated = pm.MutableData("treated", df.treated, dims="obs_id")
+    sigma = pm.HalfNormal("sigma", 1)
+    delta = pm.Cauchy("effect", alpha=0, beta=1)
+    mu = pm.Deterministic("mu", x + (delta * treated), dims="obs_id")
+    pm.Normal("y", mu=mu, sigma=sigma, observed=df.y, dims="obs_id")
+
+pm.model_to_graphviz(model)
+```
+
+## Inference
+
+```{code-cell} ipython3
+with model:
+    idata = pm.sample(random_seed=123)
+```
+
+We can see that we get no sampling warnings, and that plotting the MCMC traces shows no issues.
+
+```{code-cell} ipython3
+az.plot_trace(idata, var_names=["effect", "sigma"]);
+```
+
+We can also see that we are able to accurately recover the true discontinuity magnitude (left) and the standard deviation of the change in units between pre- and post-test (right).
+
+```{code-cell} ipython3
+az.plot_posterior(idata, var_names=["effect", "sigma"], ref_val=[treatment_effect, sd]);
+```
+
+## Counterfactual questions
+
+We can use posterior prediction to ask what would we expect to see if:
+- no units were exposed to the treatment (blue shaded region)
+- all units were exposed to the treatment (orange shaded region).
+
+```{code-cell} ipython3
+:tags: [hide-input]
+
+# MODEL EXPECTATION WITHOUT TREATMENT ------------------------------------
+# probe data
+_x = np.linspace(np.min(df.x), np.max(df.x), 500)
+_treated = np.zeros(_x.shape)
+
+# posterior prediction
+with model:
+    pm.set_data({"x": _x, "treated": _treated})
+    ppc = pm.sample_posterior_predictive(idata, var_names=["mu", "y"])
```

Comment:
Side note I hope helps with the other comments. I try to distinguish between posterior predictive and pushforward posterior variables. pushforward posterior variables are _deterministic_ transformations of the posterior parameters, posterior predictive are variables that need an extra sampling step. 

In this case, `mu = x + delta*treated`, once the posterior values are fixed, mu is also fixed. Here the `sample_posterior_predictive` is being used as a convenience function to recalculate `mu` with the modified data,
but even if `sample_...` is called, there is no sampling involved. There is sampling in computing `y` because values of `y` are draws from the normal distribution of mean `mu` and std `sigma`. Multiple calls to `sample_posterior_predictive` with different seeds will return different values of `y` but always the same values for `mu`.

Reply:
Understood. I have added a bit of text to clarify this. I would not want to change the code to manually calculate `mu` for new values of `x` and `treated` - I think the current code is very clear and convenient as you say. Hopefully this works for you?

Reply:
No need to change the code here at all, this was mostly for context